### PR TITLE
Solutions to lecture 1

### PR DESCRIPTION
**Pull request by mklinovsky**
### Solutions for Lecture 1

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture1.hs [hunk 1]
```diff
@@ -63,7 +65,8 @@ sumOfSquares x y = error "TODO!"
 
 -}
 -- DON'T FORGET TO SPECIFY THE TYPE IN HERE
-lastDigit n = error "lastDigit: Not implemented!"
+lastDigit :: Int -> Int
+lastDigit n = mod (abs n) 10
```
Comment (chshersh) [inline src/Lecture1.hs]: Absolutely correct 👍🏻 
Alternatively, you can use the infix form of a function by putting it inside backticks :slightly_smiling_face: 
```suggestion
lastDigit n = abs n `mod` 10
```

File: src/Lecture1.hs [hunk 2]
```diff
@@ -77,7 +80,11 @@ and 1 is the smallest, and 7 - 1 = 6.
 Try to use local variables (either let-in or where) to implement this
 function.
 -}
-minmax x y z = error "TODO"
+minmax :: Int -> Int -> Int -> Int
+minmax x y z =
+  let maxNum a b c = max a (max b c)
+      minNum a b c = min a (min b c)
+   in maxNum x y z - minNum x y z
```
Comment (chshersh) [inline src/Lecture1.hs]: You don't need to define `maxNum` and `minNum` as functions as the function arguments are already in scope for them
```suggestion
  let maxNum = max x (max y z)
      minNum = min x (min y z)
   in maxNum - minNum
```

File: src/Lecture1.hs [hunk 3]
```diff
@@ -94,7 +101,13 @@ start position can be considered as zero (e.g. substring from the
 first character) and negative end position should result in an empty
 string.
 -}
-subString start end str = error "TODO"
+subString :: Int -> Int -> [Char] -> [Char]
+subString start end str =
+    let newStart = if start < 0 then 0 else start
```
Comment (chshersh) [inline src/Lecture1.hs]: You can use a standard function `max` to write this shorter

```suggestion
    let newStart = max 0 start
```

File: src/Lecture1.hs [hunk 4]
```diff
@@ -104,7 +117,8 @@ and finds a sum of the numbers inside this string.
 
 The string contains only spaces and/or numbers.
 -}
-strSum str = error "TODO"
+strSum :: [Char] -> Int
+strSum str = sum (map (\x -> read x :: Int) (words str))
```
Comment (chshersh) [inline src/Lecture1.hs]: As you specify the top-level type signature for this function, the compiler would be able to figure out what type `read` function has in this case due to type inference, so you can skip specifying it in the body:

```suggestion
strSum str = sum (map read (words str))
```

File: src/Lecture1.hs [hunk 5]
```diff
@@ -119,4 +133,17 @@ and lower than 6 elements (4, 5, 6, 7, 8 and 9).
 
 🕯 HINT: Use recursion to implement this function.
 -}
-lowerAndGreater n list = error "TODO"
+lowerAndGreater :: Int -> [Int] -> [Char]
+lowerAndGreater n list =
+  let lower = show (count n list (<))
+      greater = show (count n list (>))
+   in show n ++ " is greater than " ++ greater ++ " elements and lower than " ++ lower ++ " elements"
```
Comment (chshersh) [inline src/Lecture1.hs]: That's correct! 💯 

However, with this solution, you will walk through the list twice.
You can write it more performant with only one traversal of the list 🙂
It's desirable to write functions that traverse lists single time and consume elements only once. In that case, the composition of such functions will have nice streaming and lazy properties.

**Review by chshersh — APPROVED** [review]
Congrats! The first homework is complete 👏🏼 
Hope you enjoyed the course so far. Keep up the great work, and feel free to ask any questions if you have 👌🏼

File: src/Lecture1.hs (hunk 1 above)
Comment (mklinovsky) [reply src/Lecture1.hs]: Not sure which form I like more, function inside backticks looks strange :)

File: src/Lecture1.hs (hunk 3 above)
Comment (mklinovsky) [reply src/Lecture1.hs]: That's much nicer 👍

File: src/Lecture1.hs (hunk 5 above)
Comment (mklinovsky) [reply src/Lecture1.hs]: Not only haskell is lazy, I am too :) I wrote a shorter solution and let the computer to do the extra work 😄  
Refactored, thanks for the review 👍

**Comment by chshersh** [comment]
@mklinovsky The PR looks good to me 👍🏻 
I think it's ready for merge 🚢

**Pull request merged**